### PR TITLE
Feature/train expert

### PR DESCRIPTION
**Pull request by ioneliabuzatu**
- implementation of the classifier and modular functions for training i.g. defining the number of classes
- torch_to_jax
- simulating with actions 
- cross entropy as cost function

### CODE REVIEW

File: src/load_utils.py [hunk 1]
```diff
@@ -35,11 +36,14 @@ def load_grn_jax(interactions_filename, adjacency):
...
-            adjacency = adjacency.at[regulators_nodes_ids, target_node_id].set(contributions)
+            # adjacency = adjacency.at[regulators_nodes_ids, target_node_id].set(contributions)
+            adjacency[regulators_nodes_ids, target_node_id] = contributions
 
             topo_sort_graph.add_weighted_edges_from(
                 zip(regulators_nodes_ids, repeat(target_node_id), contributions)
             )
+    # adjacency[adjacency>0] = 1
+    # adjacency[adjacency<0] = -1
     return adjacency, topo_sort_graph
```
Comment (manuel-delverme) [inline src/load_utils.py:47]: th efunction itself is called load_grn_jax, shouldn't it return a jax array rather than numpy?

File: src/models/expert/classfier_cell_state.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        x = torch.from_numpy(np.array(data, dtype=np.float32))
+        y = torch.from_numpy(np.array(self.labels_categorical[idx], dtype=np.float32))
+        del data
+        return x, y
+
+    def __len__(self):
+        return len(self.data)
+
+    def preprocess_data(self):
+        """ TODO: try TPM normalization too """
```
Comment (manuel-delverme) [inline src/models/expert/classfier_cell_state.py:201]: TPM?

File: src/models/expert/classfier_cell_state.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        self.labels_encoding, self.labels_categorical = np.unique(
+            np.concatenate([np.array([str(x)] * 10000, dtype=object) for x in range(9)], axis=0), return_inverse=True)
+
+    def __getitem__(self, idx):
+        # data = self.data[idx, :-1]
+        data = self.data[idx]
+        # indices_to_set_to_zero = np.random.permutation(self.num_genes_to_zero_for_batch_sgd)
+        # data = data[indices_to_set_to_zero]
+        # data[indices_to_set_to_zero] = 0.0
+        x = torch.from_numpy(np.array(data, dtype=np.float32))
```
Comment (manuel-delverme) [inline src/models/expert/classfier_cell_state.py:192]: do you really need from_numpy? you are casting it to numpy in the same line :)

File: train_expert.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    num_genes = 400
+    lr = 1e-3
+    epochs = 300
+    batch_size = 128
+    samples_each_type = 10000
+    checkpoint_folder = "src/models/expert/checkpoints/"
+
+
+config = config()
+experiment_buddy.register_defaults({'dataset': 'ds2'})
```
Comment (manuel-delverme) [inline train_expert.py:34]: you might also want to register config.__dict__
```
experiment_buddy.register_defaults({'dataset': 'ds2', **config.__dict__})
```

File: src/models/expert/classfier_cell_state.py (hunk 2 above)
Comment (ioneliabuzatu) [reply src/models/expert/classfier_cell_state.py:201]: 'transcript per million' is a normalization formula used specifically in single cell data, which accounts for gene length. https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/. i am curious to see if it makes any difference

File: train_expert.py (hunk 4 above)
Comment (ioneliabuzatu) [reply train_expert.py:34]: thanks!

File: src/models/expert/classfier_cell_state.py (hunk 3 above)
Comment (ioneliabuzatu) [reply src/models/expert/classfier_cell_state.py:192]: you mean like: x = torch.from_numpy(data)?

Comment (manuel-delverme) [reply src/models/expert/classfier_cell_state.py:192]: torch.tensor(data, dtype=torch.float32)

File: src/models/expert/classfier_cell_state.py (hunk 2 above)
Comment (manuel-delverme) [reply src/models/expert/classfier_cell_state.py:201]: you could put the https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/ link in the comment for us mortals :)
